src/main/java/org/prebid/server/analytics/reporter/greenbids/GreenbidsAnalyticsReporter.java

```diff
@@ -0,0 +1,323 @@
+                        .bidder(seat)
+                        .isTimeout(false)
+                        .hasBid(bid != null)
+                        .build();
+                bidders.add(bidder);
+            });
+
+            seatsWithNonBidsForImp.forEach((seat, nonBid) -> {
+                final GreenbidsBidder bidder = GreenbidsBidder.builder()
+                        .bidder(seat)
+                        .isTimeout(nonBid.getStatusCode().code == BidRejectionReason.TIMED_OUT.code)
+                        .hasBid(false)
+                        .build();
+                bidders.add(bidder);
+            });
+
+            final String adUnitCode = getAdUnitCode(imp);
+
+            return AdUnit.builder()
+                    .code(adUnitCode)
+                    .mediaTypes(mediaTypes)
+                    .bidders(bidders)
+                    .build();
+        }).toList();
+
+        return commonMessage;
+    }
+
+    private static SeatNonBid toSeatNonBid(String bidder, BidRejectionTracker bidRejectionTracker) {
+        final List<NonBid> nonBids = bidRejectionTracker.getRejectionReasons().entrySet().stream()
+                .map(entry -> NonBid.of(entry.getKey(), entry.getValue()))
+                .toList();
+
+        return SeatNonBid.of(bidder, nonBids);
+    }
+
+    public CommonMessage createBidMessage(
+            AuctionContext auctionContext,
+            BidResponse bidResponse
+    ) {
+        final List<Imp> imps = auctionContext.getBidRequest().getImp();
+
+        if (imps == null || imps.isEmpty()) {
+            throw new IllegalArgumentException("imps is null or empty");
+        }
+
+        final Long auctionTimestamp = Optional.of(auctionContext.getBidRequest())
+                .map(BidRequest::getExt)
+                .map(ExtRequest::getPrebid)
+                .map(ExtRequestPrebid::getAuctiontimestamp)
+                .orElse(null);
+
+        final long auctionElapsed = auctionTimestamp != null ? System.currentTimeMillis() - auctionTimestamp : 0L;
+
+        final Map<String, Bid> seatsWithBids = Optional.ofNullable(bidResponse.getSeatbid())
+                .orElseGet(Collections::emptyList)
+                .stream()
+                .filter(seatBid -> !seatBid.getBid().isEmpty())
+                .collect(
+                        Collectors.toMap(
+                                SeatBid::getSeat,
+                                seatBid -> seatBid.getBid().get(0),
+                                (existing, replacement) -> existing
+                        )
+                );
+
+        final List<SeatNonBid> seatNonBids = auctionContext.getBidRejectionTrackers().entrySet().stream()
+                .map(entry -> toSeatNonBid(entry.getKey(), entry.getValue()))
+                .filter(seatNonBid -> !seatNonBid.getNonBid().isEmpty())
+                .toList();
+
+        final Map<String, NonBid> seatsWithNonBids = Optional.of(seatNonBids)
```

Comment:
self-review: When I’ve been testing the AuctionContext.bidRejectionTrackers sending the [request for storedresponse](https://github.com/prebid/prebid-server-java/blob/master/sample/requests/rubicon-storedresponse.json) for the Rubicon bidder I get both

- `seatsWithNonBids`: `{rubicon=NonBid(impId=a, statusCode=NO_BID)}` extracted from `AuctionContext.bidRejectionTrackers`
- `seatsWithBids`: `{rubicon=Bid(id=1, impid=a, price=1.23)}` extracted from `AuctionContext.BidResponse.seatBid`
is it possible to have 2 responses for the one ad unit  from the same bidder in both seat no bid and seat bid or maybe I’m missing something?

Reply:
I have also asked this question to PBS Team, they have opened a ticket to investigate the issue

![Uploading image.png…]()